src/Microsoft.AspNetCore.Razor.Language/Legacy/ImplicitExpressionEditHandler.cs

```diff
@@ -216,8 +221,152 @@ private static bool IsAcceptableDeletion(Span target, SourceChange change)
         private static bool IsAcceptableInsertion(Span target, SourceChange change)
         {
             return change.IsInsert &&
-                   (IsAcceptableEndInsertion(target, change) ||
-                   IsAcceptableInnerInsertion(target, change));
+                (IsAcceptableEndInsertion(target, change) ||
+                IsAcceptableInnerInsertion(target, change));
+        }
+
+        // Internal for testing
+        internal static bool IsAcceptableInsertionInBalancedParenthesis(Span target, SourceChange change)
+        {
+            if (change.NewText.IndexOfAny(new[] { '(', ')' }) >= 0)
+            {
+                // Insertions of parenthesis aren't handled by us. If someone else wants to accept it, they can.
+                return false;
+            }
+
+            // Determine parenthesis balance to the left of the insertion point
+            var balanceCount = 0;
+            var foundInsertionPoint = false;
+            for (var i = 0; i < target.Symbols.Count; i++)
+            {
+                var currentToken = target.Symbols[i] as CSharpSymbol;
+
+                if (AtInsertionPoint(change, currentToken))
+                {
+                    if (balanceCount == 0)
+                    {
+                        // Insertion point is outside of parenthesis, i.e. inserting at the pipe: @Foo|Baz()
+                        return false;
+                    }
+
+                    foundInsertionPoint = true;
+                }
+
+                if (!TryUpdateBalanceCount(currentToken, ref balanceCount))
+                {
+                    // Couldn't update the count. This usually occurrs when we run into a ')' outside of any parenthesis.
+                    return false;
+                }
+
+                if (foundInsertionPoint && balanceCount == 0)
+                {
+                    // Once parenthesis become balanced after the insertion point we return true, no need to go further.
+                    // If they get unbalanced down the line the expression was already unbalanced to begin with and this
+                    // change happens prior to any ambiguity.
+                    return true;
+                }
+            }
+
+            // Unbalanced parenthesis
+            return false;
+        }
+
+        // Internal for testing
+        internal static bool AtInsertionPoint(SourceChange change, CSharpSymbol currentToken)
```

Comment:
Should this be thought of as "contains insertion point"? 

why is it safe to assume that the insertion point can be at the start of a token but not the end?

Reply:
The end is equivalent to the insertion point being at the next symbol.